### PR TITLE
ipsec: Simplify XFRM IN policies

### PR DESCRIPTION
**Pull request by pchaigno**
**TL;DR.** We only need one of an XFRM IN policy or an XFRM IN state to match each packet. This commit removes one superfluous XFRM IN policy and enables some additional simplification as a result.

### Explanation

What XFRM IN policy we install currently depends of whether we are running in tunneling mode and with or without endpoint routes:

In tunneling mode:

    XFRM IN policy on mark 0x200/0xf00 (for proxy)
    XFRM IN policy on mark 0xd00/0xf00 (for decrypt.)

In native routing mode with endpoint routes:

    XFRM IN policy on mark 0x200/0xf00 (for proxy)

In all cases, we also have:

    XFRM IN state on mark 0xd00/0xf00 (for decrypt.)

The two policies in tunneling mode were introduced by a9f18f36 ("datapath/linux/ipsec: Insert additional In rule when tunneling"). The additional case for endpoint routes was introduced by 3ffe49e1 ("ipsec: Fix L7 with endpoint routes"). Now, I got to wonder how 3ffe49e1 even worked as it was missing an XFRM IN policy for `0xd00` which a9f18f36 suggested was necessary.

After some local testing, it turns out that the two XFRM IN policies for tunneling mode are not required. All we need is to have either (1) an XFRM IN policy or (2) an XFRM IN state matching the packets.

Given we always have an XFRM IN state for packets coming with the decryption mark, we don't need an XFRM IN policy for that. We only need an XFRM IN policy for packets coming with the proxy mark because we don't have a state for those, rightly so as we don't want to decrypt them.

This pull request therefore removes the XFRM IN policy for decryption. It also removes any dependency on particular options: we will always install the XFRM IN policy for the proxy. It doesn't hurt to have that policy even if not required (e.g., in native routing mode without endpoint routes).

### How was this tested?

This change was tested with our Jenkins IPsec tests (including the quarantined one for VXLAN), as well as with GKE and EKS clusters of 3 nodes. In all cases, the connectivity tests were executed and L7 policies were thus covered.

### CODE REVIEW

**Review by jibi — APPROVED** [review]

**Comment by pchaigno** [comment]
I force pushed a change only to the commit description to address some feedback from Jibi. Before that, only [k8s-1.23-kernel-5.4](https://jenkins.cilium.io/job/Cilium-PR-K8s-1.23-kernel-5.4/222/) was failing, with known flake https://github.com/cilium/cilium/issues/16122. Marking ready to merge.
